<a href="https://colab.research.google.com/github/juliafoliveira-tech/Bioestat-stica/blob/main/Semin%C3%A1rio.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

* Julia Feitosa NºUSP: 16228410
* Raissa Pereira Lima da Silva NºUSP: 14605077  
* Stéfani Machado NºUSP: 15291452

* FERREIRA, Leidiane Aparecida et al. Bioestimulante e fertilizante associados ao tratamento de sementes de milho. Revista Brasileira de Sementes, v. 29, n. 2, p. 80–89, ago. 2007. Disponível em: https://www.scielo.br/j/rbs/a/VRSZ5ZJJH4YpDKjHvPLKJdt/?lang=pt. Acesso em 28 ago. 2026.

O objetivo do trabalho aqui utilizado foi avaliar o efeito do bioestimulante Stimulate® e do fertilizante líquido Cellerate®, via tratamento de sementes, seis meses antes da semeadura e na pré-semeadura, na qualidade fisiológica das sementes de um híbrido simples (GNZ 2004) e de uma linhagem (L57) e na produtividade da cultura do milho.

In [1]:
import pandas as pd

df = pd.read_excel('Seminário.xlsx')

The `df` DataFrame contains raw data from the Excel file. Let's display the head and tail of the DataFrame to better understand its structure and identify where the actual data for plotting begins.

In [ ]:
display(df.head(10))
display(df.tail())

,"Tabela 4. Porcentagem de germinação e emergência em teste de frio de plântulas de linhagem de milho oriundas de sementes tratadas com diferentes produtos em duas épocas. UFLA, Lavras, MG, 2006.",Unnamed: 1,Unnamed: 2,Unnamed: 3,Unnamed: 4
0,Produtos,Germinação,NaN,Emergência,NaN
1,NaN,6 meses antes da semeadura,Pré-semeadura,6 meses antes da semeadura,Pré-semeadura
2,Testemunha,"91,0 Aa","97,0 aA","88,5 aA","90,5 aA"
3,Stimulate,"92,0 aA","90,0 aA","92,0 aA","91,5 aA"
4,Cellerate 1,"91,5 aA","92,5 aA","89,0 aA","90,5 aA"
5,Cellerate 2,"92,0 aA","71,5 bB","87,0 aA","79,0 bB"
6,NaN,NaN,NaN,NaN,NaN
7,As médias seguidas pela mesma letra não difere...,NaN,NaN,NaN,NaN


,"Tabela 4. Porcentagem de germinação e emergência em teste de frio de plântulas de linhagem de milho oriundas de sementes tratadas com diferentes produtos em duas épocas. UFLA, Lavras, MG, 2006.",Unnamed: 1,Unnamed: 2,Unnamed: 3,Unnamed: 4
3,Stimulate,"92,0 aA","90,0 aA","92,0 aA","91,5 aA"
4,Cellerate 1,"91,5 aA","92,5 aA","89,0 aA","90,5 aA"
5,Cellerate 2,"92,0 aA","71,5 bB","87,0 aA","79,0 bB"
6,NaN,NaN,NaN,NaN,NaN
7,As médias seguidas pela mesma letra não difere...,NaN,NaN,NaN,NaN


Based on the `df.head(10)` output, the actual data for plotting is in rows 2 to 5, and the relevant columns are the product names, and the germination percentages for '6 meses antes da semeadura' and 'Pré-semeadura'.

I'll now perform the following steps:
1. Select the relevant rows (index 2 to 5) and columns.
2. Rename the columns to more descriptive names.
3. Clean the percentage values by removing text like 'aA' or 'bB' and converting them to a numeric type.
4. Melt the DataFrame to a long format to prepare for plotting, where each row will represent a product, a time period, and its corresponding germination percentage.

In [ ]:
import re

# Extract relevant rows and columns
df_germination = df.iloc[2:6, [0, 1, 2]].copy()

# Rename columns for clarity
df_germination.columns = ['Produto', '6 meses antes da semeadura', 'Pré-semeadura']

# Function to clean percentage strings and convert to float using regex
def clean_percentage(value):
    if isinstance(value, str):
        # Replace comma with dot for decimal parsing
        value = value.replace(',', '.')
        # Use regex to find the first numeric pattern (integer or float)
        match = re.search(r'(\d+(\.\d+)?)', value)
        if match:
            return float(match.group(1))
    # If it's already a float or no match, return as is (or convert if needed)
    return float(value)

# Apply cleaning to percentage columns
df_germination['6 meses antes da semeadura'] = df_germination['6 meses antes da semeadura'].apply(clean_percentage)
df_germination['Pré-semeadura'] = df_germination['Pré-semeadura'].apply(clean_percentage)

# Melt the DataFrame to long format for easier plotting
df_germination_melted = df_germination.melt(
    id_vars=['Produto'],
    value_vars=['6 meses antes da semeadura', 'Pré-semeadura'],
    var_name='Período',
    value_name='Porcentagem de Germinação'
)

display(df_germination_melted)

,Produto,Período,Porcentagem de Germinação
0,Testemunha,6 meses antes da semeadura,91.0
1,Stimulate,6 meses antes da semeadura,92.0
2,Cellerate 1,6 meses antes da semeadura,91.5
3,Cellerate 2,6 meses antes da semeadura,92.0
4,Testemunha,Pré-semeadura,97.0
5,Stimulate,Pré-semeadura,90.0
6,Cellerate 1,Pré-semeadura,92.5
7,Cellerate 2,Pré-semeadura,71.5


Now that the data is cleaned and reshaped into `df_germination_melted`, I'll create the bar chart using Altair to visualize the germination percentages. The chart will compare '6 meses antes da semeadura' and 'Pré-semeadura' for each product.

In [ ]:
import altair as alt

chart = alt.Chart(df_germination_melted).mark_bar().encode(
    x=alt.X('Produto:N', axis=alt.Axis(title='Produto')),
    y=alt.Y('Porcentagem de Germinação:Q', axis=alt.Axis(title='Porcentagem de Germinação', format='.1f', labelExpr="datum.value + '%'")),
    color=alt.Color('Período:N', title='Período'),
    column=alt.Column('Período:N', header=alt.Header(titleOrient="bottom", labelOrient="bottom")),
    tooltip=[
        alt.Tooltip('Produto:N', title='Produto'),
        alt.Tooltip('Período:N', title='Período'),
        alt.Tooltip('Porcentagem de Germinação:Q', title='Germinação', format='.1f')
    ]
).properties(
    title='Porcentagem de Germinação por Produto e Período',
    width=200, # Adjust width of each column
    height=300 # Adjust height of the chart
).interactive()

chart

alt.Chart(...)

Now, let's create a scatter plot to visualize the germination percentages, comparing each product across the two periods ('6 meses antes da semeadura' and 'Pré-semeadura'). This will allow for a direct comparison of the germination performance.

In [ ]:
import altair as alt

scatter_chart = alt.Chart(df_germination_melted).mark_circle(size=100).encode(
    x=alt.X('Produto:N', axis=alt.Axis(title='Produto')),
    y=alt.Y('Porcentagem de Germinação:Q', axis=alt.Axis(title='Porcentagem de Germinação', format='.1f', labelExpr="datum.value + '%'")),
    color=alt.Color('Período:N', title='Período'),
    tooltip=[
        alt.Tooltip('Produto:N', title='Produto'),
        alt.Tooltip('Período:N', title='Período'),
        alt.Tooltip('Porcentagem de Germinação:Q', title='Germinação', format='.1f')
    ]
).properties(
    title='Germinação por Produto e Período (Gráfico de Dispersão)',
    width=600,
    height=400
).interactive()

scatter_chart

alt.Chart(...)

Now, let's create a scatter plot for 'Emergência' (Emergence), following similar steps as for germination:
1. Extract the relevant rows (index 2 to 5) and columns (0, 3, 4) that contain the product names and emergence percentages.
2. Rename the columns appropriately.
3. Clean the percentage values using the already defined `clean_percentage` function.
4. Melt the DataFrame to a long format to prepare it for plotting.
5. Finally, create the Altair scatter plot.

In [ ]:
import re
import altair as alt

# Extract relevant rows and columns for emergence data
df_emergence = df.iloc[2:6, [0, 3, 4]].copy()

# Rename columns for clarity
df_emergence.columns = ['Produto', '6 meses antes da semeadura', 'Pré-semeadura']

# The clean_percentage function is already defined from previous steps
# Apply cleaning to percentage columns
df_emergence['6 meses antes da semeadura'] = df_emergence['6 meses antes da semeadura'].apply(clean_percentage)
df_emergence['Pré-semeadura'] = df_emergence['Pré-semeadura'].apply(clean_percentage)

# Melt the DataFrame to long format for easier plotting
df_emergence_melted = df_emergence.melt(
    id_vars=['Produto'],
    value_vars=['6 meses antes da semeadura', 'Pré-semeadura'],
    var_name='Período',
    value_name='Porcentagem de Emergência'
)

display(df_emergence_melted)

,Produto,Período,Porcentagem de Emergência
0,Testemunha,6 meses antes da semeadura,88.5
1,Stimulate,6 meses antes da semeadura,92.0
2,Cellerate 1,6 meses antes da semeadura,89.0
3,Cellerate 2,6 meses antes da semeadura,87.0
4,Testemunha,Pré-semeadura,90.5
5,Stimulate,Pré-semeadura,91.5
6,Cellerate 1,Pré-semeadura,90.5
7,Cellerate 2,Pré-semeadura,79.0


In [ ]:
scatter_emergence_chart = alt.Chart(df_emergence_melted).mark_circle(size=100).encode(
    x=alt.X('Produto:N', axis=alt.Axis(title='Produto')),
    y=alt.Y('Porcentagem de Emergência:Q', axis=alt.Axis(title='Porcentagem de Emergência', format='.1f', labelExpr="datum.value + '%'")),
    color=alt.Color('Período:N', title='Período'),
    tooltip=[
        alt.Tooltip('Produto:N', title='Produto'),
        alt.Tooltip('Período:N', title='Período'),
        alt.Tooltip('Porcentagem de Emergência:Q', title='Emergência', format='.1f')
    ]
).properties(
    title='Emergência por Produto e Período (Gráfico de Dispersão)',
    width=600,
    height=400
).interactive()

scatter_emergence_chart

alt.Chart(...)

Agora, vamos criar um boxplot para a taxa de germinação, que mostrará a distribuição das porcentagens de germinação para cada produto nos dois períodos.

In [ ]:
import altair as alt

boxplot_germination_chart = alt.Chart(df_germination_melted).mark_boxplot().encode(
    x=alt.X('Produto:N', axis=alt.Axis(title='Produto')),
    y=alt.Y('Porcentagem de Germinação:Q', axis=alt.Axis(title='Porcentagem de Germinação', format='.1f', labelExpr="datum.value + '%'")),
    color=alt.Color('Período:N', title='Período'),
    tooltip=[
        alt.Tooltip('Produto:N', title='Produto'),
        alt.Tooltip('Período:N', title='Período'),
        alt.Tooltip('Porcentagem de Germinação:Q', title='Germinação', format='.1f')
    ]
).properties(
    title='Boxplot da Porcentagem de Germinação por Produto e Período',
    width=600,
    height=400
).interactive()

boxplot_germination_chart

alt.Chart(...)

In [ ]:
primeiro_quartil_germinacao = df_germination_melted['Porcentagem de Germinação'].quantile(0.25)
print(f"O primeiro quartil para a taxa de germinação é: {primeiro_quartil_germinacao:.2f}%")

segundo_quartil_germinacao = df_germination_melted['Porcentagem de Germinação'].quantile(0.50)
print(f"O segundo quartil (mediana) para a taxa de germinação é: {segundo_quartil_germinacao:.2f}%")

terceiro_quartil_germinacao = df_germination_melted['Porcentagem de Germinação'].quantile(0.75)
print(f"O terceiro quartil para a taxa de germinação é: {terceiro_quartil_germinacao:.2f}%")

O primeiro quartil para a taxa de germinação é: 90.75%
O segundo quartil (mediana) para a taxa de germinação é: 91.75%
O terceiro quartil para a taxa de germinação é: 92.12%


* O primeiro quartil indica que 25% das sementes germinam a uma taxa de 90,75% ou menos.
* O segundo quartil indica que 50% das sementes germinam em uma taxa de 91,75%, ou seja, é a mediana o valor central da taxa de germinação de sementes.
*O terceiro quartil indica que 25% das sementes germinam a uma taxa de 92,12% ou mais.

In [ ]:
intervalo_interquartil_germinacao = terceiro_quartil_germinacao - primeiro_quartil_germinacao
print(f"O Intervalo Interquartil (IQR) para a taxa de germinação é: {intervalo_interquartil_germinacao:.2f}%")

O Intervalo Interquartil (IQR) para a taxa de germinação é: 1.38%


In [ ]:
primeiro_quartil_emergencia = df_emergence_melted['Porcentagem de Emergência'].quantile(0.25)
print(f"O primeiro quartil para a taxa de emergência é: {primeiro_quartil_emergencia:.2f}%")

O primeiro quartil para a taxa de emergência é: 88.12%


In [ ]:
import statsmodels.api as sm
from statsmodels.formula.api import ols

# Análise ANOVA para Germinação (sem termo de interação)
# Removendo o termo de interação C(Produto):C(Período) devido à limitação de dados (apenas 1 observação por célula)
model_germinacao = ols('Q("Porcentagem de Germinação") ~ C(Produto) + C(Período)', data=df_germination_melted).fit()
anova_table_germinacao = sm.stats.anova_lm(model_germinacao, typ=2)

print("ANOVA para Porcentagem de Germinação (sem interação):")
print(anova_table_germinacao)

print("\n---\n")

# Análise ANOVA para Emergência (sem termo de interação)
# Removendo o termo de interação C(Produto):C(Período) devido à limitação de dados (apenas 1 observação por célula)
model_emergencia = ols('Q("Porcentagem de Emergência") ~ C(Produto) + C(Período)', data=df_emergence_melted).fit()
anova_table_emergencia = sm.stats.anova_lm(model_emergencia, typ=2)

print("ANOVA para Porcentagem de Emergência (sem interação):")
print(anova_table_emergencia)

ANOVA para Porcentagem de Germinação (sem interação):
               sum_sq   df         F    PR(>F)
C(Produto)  177.34375  3.0  0.884094  0.539139
C(Período)   30.03125  1.0  0.449135  0.550720
Residual    200.59375  3.0       NaN       NaN

---

ANOVA para Porcentagem de Emergência (sem interação):
            sum_sq   df         F    PR(>F)
C(Produto)  86.750  3.0  2.700389  0.218112
C(Período)   3.125  1.0  0.291829  0.626577
Residual    32.125  3.0       NaN       NaN


## Produtos Mais Eficientes: Análise Resumida

In [ ]:
# Calcular a média da Porcentagem de Germinação por Produto
media_germinacao_por_produto = df_germination_melted.groupby('Produto')['Porcentagem de Germinação'].mean().reset_index()
media_germinacao_por_produto.rename(columns={'Porcentagem de Germinação': 'Média de Germinação (%)'}, inplace=True)

# Calcular a média da Porcentagem de Emergência por Produto
media_emergencia_por_produto = df_emergence_melted.groupby('Produto')['Porcentagem de Emergência'].mean().reset_index()
media_emergencia_por_produto.rename(columns={'Porcentagem de Emergência': 'Média de Emergência (%)'}, inplace=True)

# Unir as duas tabelas de médias
tabela_eficiencia_produtos = pd.merge(media_germinacao_por_produto, media_emergencia_por_produto, on='Produto')

print("Tabela 1: Desempenho Médio de Germinação e Emergência por Produto")
display(tabela_eficiencia_produtos)

print("\n--- Análise da Tabela 1 ---")
produto_maior_germinacao = tabela_eficiencia_produtos.loc[tabela_eficiencia_produtos['Média de Germinação (%)'].idxmax()]
produto_maior_emergencia = tabela_eficiencia_produtos.loc[tabela_eficiencia_produtos['Média de Emergência (%)'].idxmax()]

print(f"O produto com a maior média de germinação é: {produto_maior_germinacao['Produto']} ({produto_maior_germinacao['Média de Germinação (%)']:.2f}%)")
print(f"O produto com a maior média de emergência é: {produto_maior_emergencia['Produto']} ({produto_maior_emergencia['Média de Emergência (%)']:.2f}%)\n")


Tabela 1: Desempenho Médio de Germinação e Emergência por Produto


,Produto,Média de Germinação (%),Média de Emergência (%)
0,Cellerate 1,92.00,89.75
1,Cellerate 2,81.75,83.00
2,Stimulate,91.00,91.75
3,Testemunha,94.00,89.50



--- Análise da Tabela 1 ---
O produto com a maior média de germinação é: Testemunha (94.00%)
O produto com a maior média de emergência é: Stimulate (91.75%)



In [ ]:
# Criar DataFrame para resumir os resultados da ANOVA
resumo_anova = pd.DataFrame({
    'Análise': ['Germinação', 'Germinação', 'Emergência', 'Emergência'],
    'Fator': ['Produto', 'Período', 'Produto', 'Período'],
    'p-valor (PR(>F))': [
        anova_table_germinacao.loc['C(Produto)', 'PR(>F)'],
        anova_table_germinacao.loc['C(Período)', 'PR(>F)'],
        anova_table_emergencia.loc['C(Produto)', 'PR(>F)'],
        anova_table_emergencia.loc['C(Período)', 'PR(>F)']
    ],
    'Significativo (p<0.05)': [
        anova_table_germinacao.loc['C(Produto)', 'PR(>F)'] < 0.05,
        anova_table_germinacao.loc['C(Período)', 'PR(>F)'] < 0.05,
        anova_table_emergencia.loc['C(Produto)', 'PR(>F)'] < 0.05,
        anova_table_emergencia.loc['C(Período)', 'PR(>F)'] < 0.05
    ]
})

print("Tabela 2: Resumo dos Resultados da Análise ANOVA (sem interação)")
display(resumo_anova)

print("\n--- Análise da Tabela 2 ---")
print("Nenhuma das análises de ANOVA (para Germinação e Emergência, considerando 'Produto' e 'Período' separadamente) mostrou um efeito estatisticamente significativo (p-valor < 0.05).")
print("Isso sugere que, com base nos dados disponíveis e no modelo ANOVA utilizado, não há diferenças estatisticamente comprovadas entre os produtos ou períodos para as taxas de germinação e emergência.")
print("É importante lembrar que esta conclusão é limitada pelo tamanho da amostra e pela ausência do termo de interação no modelo de ANOVA, devido à estrutura dos dados.")


Tabela 2: Resumo dos Resultados da Análise ANOVA (sem interação)


,Análise,Fator,p-valor (PR(>F)),Significativo (p<0.05)
0,Germinação,Produto,0.539139,False
1,Germinação,Período,0.550720,False
2,Emergência,Produto,0.218112,False
3,Emergência,Período,0.626577,False



--- Análise da Tabela 2 ---
Nenhuma das análises de ANOVA (para Germinação e Emergência, considerando 'Produto' e 'Período' separadamente) mostrou um efeito estatisticamente significativo (p-valor < 0.05).
Isso sugere que, com base nos dados disponíveis e no modelo ANOVA utilizado, não há diferenças estatisticamente comprovadas entre os produtos ou períodos para as taxas de germinação e emergência.
É importante lembrar que esta conclusão é limitada pelo tamanho da amostra e pela ausência do termo de interação no modelo de ANOVA, devido à estrutura dos dados.


## Correlação de Pearson entre Germinação e Emergência por Produto

In [ ]:
import pandas as pd

# Calcular a Correlação de Pearson entre as médias de germinação e emergência
correlacao_pearson = tabela_eficiencia_produtos['Média de Germinação (%)'].corr(tabela_eficiencia_produtos['Média de Emergência (%)'], method='pearson')

print(f"A Correlação de Pearson entre a Média de Germinação e a Média de Emergência por Produto é: {correlacao_pearson:.2f}")

if correlacao_pearson > 0.7:
    print("Isso indica uma forte correlação positiva: produtos com maior germinação tendem a ter maior emergência.")
elif correlacao_pearson < -0.7:
    print("Isso indica uma forte correlação negativa: produtos com maior germinação tendem a ter menor emergência.")
elif correlacao_pearson >= 0.3 and correlacao_pearson <= 0.7:
    print("Isso indica uma correlação positiva moderada: existe uma tendência de produtos com maior germinação terem maior emergência, mas a relação não é muito forte.")
elif correlacao_pearson >= -0.7 and correlacao_pearson <= -0.3:
    print("Isso indica uma correlação negativa moderada: existe uma tendência de produtos com maior germinação terem menor emergência, mas a relação não é muito forte.")
else:
    print("Isso indica uma correlação fraca ou inexistente: não há uma relação linear clara entre a germinação e a emergência dos produtos com base nas médias.")

A Correlação de Pearson entre a Média de Germinação e a Média de Emergência por Produto é: 0.89
Isso indica uma forte correlação positiva: produtos com maior germinação tendem a ter maior emergência.


* Há uma relação direta e consideravelmente forte entre o desempenho de germinação e o desempenho de emergência para esses produtos, com base nas médias observadas.



## Gráfico de Dispersão com Regressão Linear: Germinação vs. Emergência

In [ ]:
import altair as alt

# Criar o gráfico de dispersão
chart = alt.Chart(tabela_eficiencia_produtos).mark_circle(size=100).encode(
    x=alt.X('Média de Germinação (%):Q', axis=alt.Axis(title='Média de Germinação (%)')),
    y=alt.Y('Média de Emergência (%):Q', axis=alt.Axis(title='Média de Emergência (%)')),
    color=alt.Color('Produto:N', title='Produto'),
    tooltip=[
        alt.Tooltip('Produto:N', title='Produto'),
        alt.Tooltip('Média de Germinação (%):Q', title='Germinação Média', format='.2f'),
        alt.Tooltip('Média de Emergência (%):Q', title='Emergência Média', format='.2f')
    ]
).properties(
    title='Correlação entre Média de Germinação e Emergência por Produto'
)

# Adicionar a reta de regressão linear
reg_line = chart.transform_regression(
    'Média de Germinação (%)', 'Média de Emergência (%)'
).mark_line(color='firebrick')

# Combinar o gráfico de dispersão com a linha de regressão
final_chart = chart + reg_line

final_chart

alt.LayerChart(...)

* Análise do Gráfico:

Cada ponto no gráfico representa um produto, com sua posição determinada pela 'Média de Germinação (%)' no eixo X e pela 'Média de Emergência (%)' no eixo Y.
As cores distinguem os diferentes produtos, permitindo identificar o desempenho individual de cada um.
A reta de regressão linear (em vermelho) ilustra a tendência geral da relação entre as duas variáveis. A inclinação ascendente da reta confirma a correlação positiva forte que calculamos anteriormente (0.89).
Podemos observar que os pontos estão relativamente próximos à reta de regressão, o que visualmente reforça a ideia de que produtos com maior média de germinação também tendem a ter maior média de emergência.
Este gráfico oferece uma representação visual clara da forte associação linear entre as médias de germinação e emergência dos produtos.